overlord/devicestate/handlers_systems.go

```diff
@@ -0,0 +1,374 @@
+		}
+	}
+	if err := s.Err(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *DeviceManager) doCreateRecoverySystem(t *state.Task, _ *tomb.Tomb) (err error) {
+	if release.OnClassic {
+		// TODO: this may need to be lifted in the future
+		return fmt.Errorf("cannot run update gadget assets task on a classic system")
+	}
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	remodelCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return err
+	}
+	model := remodelCtx.Model()
+
+	setup, err := taskRecoverySystemSetup(t)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot obtain recovery system setup information")
+	}
+	label := setup.Label
+	systemDirectory := setup.Directory
+
+	// get all infos
+	infoGetter := func(name string) (*snap.Info, bool, error) {
+		// snap may be present in the system in which case info comes
+		// from snapstate
+		info, err := snapstate.CurrentInfo(st, name)
+		if err == nil {
+			hash, _, err := asserts.SnapFileSHA3_384(info.MountFile())
+			if err != nil {
+				return nil, true, fmt.Errorf("cannot compute SHA3 of snap file: %v", err)
+			}
+			info.Sha3_384 = hash
+			return info, true, nil
+		}
+		if _, ok := err.(*snap.NotInstalledError); !ok {
+			return nil, false, err
+		}
+		logger.Debugf("requested info for not yet installed snap %q", name)
+		// TODO: handle remodel case in which snap may not be installed
+		// yet, and thus we need to pull info from snapsup of relevant
+		// download tasks
+		return nil, false, fmt.Errorf("not implemented")
+	}
+
+	observeSnapFileWrite := func(recoverySystemDir, where string) error {
+		if recoverySystemDir != systemDirectory {
+			return fmt.Errorf("internal error: unexpected recovery system path %q", recoverySystemDir)
+		}
+		// track all the files, both asserted shared snaps and private
+		// ones
+		return logNewSystemSnapFile(filepath.Join(recoverySystemDir, "snapd-new-file-log"), where)
+	}
+
+	db := assertstate.DB(st)
+	defer func() {
+		if err == nil {
+			return
+		}
+		if err := purgeNewSystemSnapFiles(filepath.Join(systemDirectory, "snapd-new-file-log")); err != nil {
+			logger.Noticef("when removing seed files: %v", err)
+		}
+		// this is ok, as before the change with this task was created,
+		// we checked that the system directory did not exist; it may
+		// exist now if one of the post-create steps failed, or the the
+		// task is being re-run after a reboot and creating a system
+		// failed
+		if err := os.RemoveAll(systemDirectory); err != nil && !os.IsNotExist(err) {
+			logger.Noticef("when removing recovery system %q: %v", label, err)
+		}
+		if err := boot.DropRecoverySystem(remodelCtx, label); err != nil {
+			logger.Noticef("when dropping the recovery system %q: %v", label, err)
+		}
+		// we could have reentered the task after a reboot, but the
+		// state was set up sufficiently such that the system was
+		// actually tried and ended up in the tried systems list, which
+		// we should reset now
```

Comment:
I don't think we are in hurry to support that but we should leave a TODO about detecting when we are half-way/almost done and, resent and/or continue instead